### ISSUE TITLE
Ideas Regarding Servers and Latencies

### ISSUE THREAD

**Issue by moriel5** [TeamPiped/Piped#224]
Since I have noticed this with both Invidious and Piped, I was thinking.

Latencies between my country and others, frankly speaking, bad (55-65ms is considered good, with the average being 75-135ms), however latencies within the country can go as low as 5ms (I generally see 7-10ms to CloudFlare's local CDN with  their DNS, with my ISP's DNS's latencies being ~14-24ms).

Case in point, regardless of whether I use 1.1.1.1 or my ISP's DNS, the latency to the CDN that locally hosts YouTube is ~44-48ms), to Snopyta's Invidious instance, it is about 54-58ms, however to kavin.rocks it is about 64-65ms (used to be ~75-80ms). To alefvanoon.xyz it is either ~59-61ms or ~70-71ms (keeps on jumping between the two).

Currently, I believe that the content is served like this, when not proxied (which confuses me as to how it works when proxied):

1. Request: User > Invidious/Piped instance > CDN in the area of the server where the instance runs on > YouTube
2. Content: YouTube > CDN in the area of the server where the instance runs on > Invidious/Piped instance > User

If so, I was wondering whether there is a way to have the content go through the CDN that is around the user's area, something like this:

1. Base implementation: User request > Piped instance > Piped website is served from the instance's server to the user
2. Video/Audio Metadata: User request > Invidious/Piped instance > CDN in the area of the server where the instance runs on > YouTube > metadata content via CDN in the area of the server where the instance runs on > Invidious/Piped instance > User
3. Actual Streams + Geolocation Metadata: User request > CDN in the user's area > YouTube >  content via CDN in user's area > YouTube.

In theory, this may be possible, since the NewPipe Android app connects via the local CDN (since it runs locally), though a website and an app are built the same way, this may prove to be challenging.

I may be wrong, and there could also be better approaches to this, which I would also like to hear (disclaimer, I am not one of the developers, just a user here).

**Comment by FireMasterK** [comment]
I feel you may not be understanding the internal working of Piped, let me explain some things first.

Website: It is downloaded once, and served by the browser through the use of service workers. No requests would be made apart from update checking every time the page is opened.

Once the page is loded locally, an API request is made to `pipedapi.kavin.rocks` and receives a response in json and is then rendered locally in the browser.

So, here's why it is efficient for cached queries:

In the case of an already cached query, this would be the path since Cloudflare has already cached the query.
Browser -> Cloudflare 

In case of an uncached query, here is the process:
Browser -> Cloudflare -> PipedAPI Server -> YouTube

Now, the response is cached and used for future responses.

The same goes for images.

In Invidious, here's how the path would be for a cached query.

Browser -> Cloudflare -> Invidious

Now, here's the same for an uncached query:

Browser -> Cloudflare -> Invidious -> YouTube

For images, however, they can be cached just like Piped.

This is how Piped avoids 100-200ms of delay in the case of a cached response.

Regarding why 1.1.1.1 has 11 ms of latency and kavin.rocks has 60 ms of latency, its because of how ISPs peer with each and cloudflare.

Since not every ISP peers with Cloudflare, it costs Cloudflare money to send traffic through non-peered routes.

In short, 1.1.1.1 has lower latency because it is running on the enterprise plan, whereas kavin.rocks is running on the free plan.

**Comment by moriel5** [comment]
Thanks, it would appear that I really did not understand this.
However, in that case, it sounds as though my queries are not being cached, as I would expect, after the initial load (I have cache and cookies cleared upon session close, but not blocked), for latencies to drop to the same level as on the YouTube website (for video contents, obviously), however I consistently get slower buffering with Piped (equivalent to what the initial latency would suggest).

**Comment by FireMasterK** [comment]
Currently, there's a plaguing change on YouTube's server that has been causing the actual video streaming to be really slow. It happens every 4-5 videos from my experience, and will not affect music video content.

Regarding cached queries (loading metadata, comments, thumbnails), they should perform just as quickly as they can be cached and served at Cloudflare itself.

There are issues tracking this in other projects too:
https://github.com/TeamNewPipe/NewPipe/issues/6510
https://github.com/ytdl-org/youtube-dl/issues/29326
https://github.com/iv-org/invidious/issues/2194

**Comment by moriel5** [comment]
Strange, as I had not witnessed this at all, and what I had said in my previous comment is, and has been for months, consistent for me.

**Comment by FireMasterK** [comment]
Strange! That should not happen at all!

Could you send a screenshot of what you see in the network tab's timings/waterfall in the webtools?
This is how it would look on chromium
![image](https://user-images.githubusercontent.com/20838718/123697107-24351380-d87a-11eb-900b-7d38296cc3e8.png)

This is how it would look on firefox
![image](https://user-images.githubusercontent.com/20838718/123697216-4333a580-d87a-11eb-9168-c12adf9115c2.png)

**Comment by FireMasterK** [comment]
A tip that I've personally used to have a better experience:
- Set a default quality/resolution in preferences
- Increase the buffering goal to a larger number (I personally use 120).

**Comment by FireMasterK** [comment]
Closing this as there seems to be nothing actionable on my side.

**Comment by moriel5** [comment]
Sorry about the lack of screenshot, I had clean forgotten. I'll try getting to it, if not in the next 10 hours, by the end of Sunday.

And interestingly, CloudTube ([tube.cadence.moe](tube.cadence.moe)), despite the website itself getting a really high ping (I forget the exact value, however it was a triple digit figure), playback buffering was flawless (just as good as the official YouTube frontend), so perhaps something could be learned from there?

**Comment by FireMasterK** [comment]
> so perhaps something could be learned from there?

They do not proxy video, ie. Google gets your IP. It also doesn't use DASH for streaming.

**Comment by moriel5** [comment]
I take it then that Piped does proxy videos?

Regarding DASH, that is quite the surprise, I was under the impression that they do, since the streaming options go up to 4K there.

**Comment by FireMasterK** [comment]
> I take it then that Piped does proxy videos?

Yes, that's correct.

> Regarding DASH, that is quite the surprise, I was under the impression that they do, since the streaming options go up to 4K there.

They just play the audio and video together, you would hence notice the audio and video going desync after a few seconds.

Using DASH would combine the audio and video properly, like how it's done in Piped and Invidious.

**Comment by moriel5** [comment]
Thanks, that would explain why I had better buffering results with different Invidious instances than with Piped, since I usually disable the proxying (I fully understand the consequences), though it falls to explain why I had the same results with your Invidious instance (with proxying off) as with Piped.

Regarding DASH, hmm... that is quite surprising. I'll ask the CloudTube developer about the lack of support, since I do not see any privacy issues with supporting DASH (though CloudTube process itself upon privacy, so it is surprising that it does not proxy the videos by default).

I take that you do not intend to add an option to Piped to disable proxying the videos, right?
